subsys/bluetooth/controller/ll_sw/ull.c

```diff
@@ -2072,13 +2072,39 @@ void ull_prepare_dequeue(uint8_t caller_id)
 	void *param_resume_head = NULL;
 	void *param_resume_next = NULL;
 	struct lll_event *next;
+	uint8_t loop;
+
+	/* Development assertion check to ensure the below loop processing
+	 * has a limit.
+	 *
+	 * We cannot use the prepapre queue allocated size as the prepare
+	 * queue does get new elements enqueued in the loop below. Hence,
+	 * say for now we have:
+	 * 2 ACL
+	 * 1 CIG
+	 * 2 continuous scanner
+	 * 1 advertiser
+	 *
+	 * of these:
+	 * 2 continuous scanner, and
+	 * 1 advertiser
+	 *
+	 * overlap, needing 9 iterations; lets check that we do not exceed 10
+	 * iterations.
+	 */
+	loop = 10U;
```

Comment:
What if we have 10 ACL connections? 
Or 16 advertising sets? 

Is the (constant) limit on this loop an upper limit that always works?

Reply:
i think my explanation is incorrect, it is only the 2 scanner and 1 advertiser (directed adv) that get enqueue back, in the which case the iterations should always max at 10:

Already in queue max 7:
- 2 continuous scan prepare in queue (1M and Coded PHY)
- 2 continuous scan resume in queue (1M and Coded PHY)
- 1 directed adv prepare
- 1 directed adv resume
- 1 any other role with time reservation

The loop removes the following 3 in 7 iterations:
- 1 scan prepare (1M)
- 1 scan prepare (Coded PHY)
- 1 directed adv prepare

and enqueues the following in these 7 iterations:
- 1 scan resume (1M)
- 1 scan resume (Coded PHY)
- 1 directed adv resume

Hence, it should be 7 + 3 iterations max.

Reply:
Ah, so it's not related to any configurable number, but rather what is theoretically possible to do in parallel. 

Would the loop value then not depend on which features we support? E.g. we are only a broadcaster, then the scanning parts can be removed and the loop value could be 3? Or 4 as peripheral? and 6 as observer and 7 as central?

Reply:
Yes, the loop can be dependent on feature support. But for now the pipeline size is not assigned based on feature support but set to max of 7 here: https://github.com/zephyrproject-rtos/zephyr/blob/42fade12c8639dfeb3258896421068f6fbc9cbed/subsys/bluetooth/controller/ll_sw/lll.h#L17

May be I can use `(EVENT_PIPELINE_MAX + 3U)`  for the check?

Reply:
> May be I can use (EVENT_PIPELINE_MAX + 3U) for the check?

I think that would make sense. IMO when values are relative to other values (like the case here), it makes sense to use the "other" value as a base for the value you are stating here. 

So if `loop = 10 == EVENT_PIPELINE_MAX + 3`, then I'd prefer to use that macro. 

Then if we at some point can reduce `EVENT_PIPELINE_MAX` based on enabled features, the `loop` variable will automatically decrease as well